### PR TITLE
Remove `PartialOrd`, `Ord` from `DefId`

### PR DESCRIPTION
**Pull request by pierwill**
Part of work on #90317.

### CODE REVIEW

**Comment by pierwill** [comment]
This work is going to require changes to chalk.

```
error[E0277]: the trait bound `rustc_span::def_id::DefId: Ord` is not satisfied
   --> compiler/rustc_middle/src/traits/chalk.rs:77:5
    |
77  |     type DefId = DefId;
    |     ^^^^^^^^^^^^^^^^^^^ the trait `Ord` is not implemented for `rustc_span::def_id::DefId`
    |
note: required by a bound in `chalk_ir::interner::Interner::DefId`
   --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/chalk-ir-0.55.0/src/interner.rs:191:37
    |
191 |     type DefId: Debug + Copy + Eq + Ord + Hash;
    |                                     ^^^ required by this bound in `chalk_ir::interner::Interner::DefId`

error[E0277]: the trait bound `adt::AdtDef: Ord` is not satisfied
   --> compiler/rustc_middle/src/traits/chalk.rs:78:5
    |
78  |     type InternedAdtId = &'tcx AdtDef;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Ord` is not implemented for `adt::AdtDef`
    |
    = note: required because of the requirements on the impl of `Ord` for `&'tcx adt::AdtDef`
note: required by a bound in `chalk_ir::interner::Interner::InternedAdtId`
   --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/chalk-ir-0.55.0/src/interner.rs:194:45
    |
194 |     type InternedAdtId: Debug + Copy + Eq + Ord + Hash;
    |                                             ^^^ required by this bound in `chalk_ir::interner::Interner::InternedAdtId`
```

File: Cargo.lock [hunk 1]
```diff
@@ -492,9 +492,8 @@ checksum = "baf1de4339761588bc0619e3cbc0120ee582ebb74b53b4efbf79117bd2da40fd"
 
 [[package]]
 name = "chalk-derive"
-version = "0.75.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d54e3b5f9e3425e6b119ff07568d8d006bfa5a8d6f78a9cbc3530b1e962e316c"
+version = "0.76.0-dev.0"
+source = "git+https://github.com/rust-lang/chalk?branch=master#f518c7c7ed723dd14486bc966e2474e46fe04269"
```
Comment (pierwill) [inline Cargo.lock:496]: This is temporary. Using this commit until a new chalk release that includes https://github.com/rust-lang/chalk/pull/740.

**Comment by pierwill** [comment]
> This work might require changes to chalk.
> 
> [...]
(https://github.com/rust-lang/rust/pull/90749#issuecomment-964699542)

Fixed in https://github.com/rust-lang/chalk/pull/740.

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 2]
```diff
@@ -1076,18 +1078,9 @@ pub trait PrettyPrinter<'tcx>:
...
             predicates.auto_traits().map(|did| (self.tcx().def_path_str(did), did)).collect();
 
-        // The auto traits come ordered by `DefPathHash`. While
-        // `DefPathHash` is *stable* in the sense that it depends on
-        // neither the host nor the phase of the moon, it depends
-        // "pseudorandomly" on the compiler version and the target.
-        //
-        // To avoid that causing instabilities in compiletest
-        // output, sort the auto-traits alphabetically.
-        auto_traits.sort();
```
Comment (pierwill) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1089]: This is probably not the best way to handle this...

File: compiler/rustc_middle/src/query/mod.rs [hunk 3]
```diff
@@ -1077,7 +1077,7 @@ rustc_queries! {
     }
 
     /// Return all `impl` blocks in the current crate.
-    query all_local_trait_impls(_: ()) -> &'tcx BTreeMap<DefId, Vec<LocalDefId>> {
+    query all_local_trait_impls(_: ()) -> &'tcx FxHashMap<DefId, Vec<LocalDefId>> {
```
Comment (bjorn3) [inline compiler/rustc_middle/src/query/mod.rs:1080]: Is it possible iteration order matters here?

File: compiler/rustc_middle/src/ty/mod.rs [hunk 4]
```diff
@@ -419,17 +417,13 @@ impl<'tcx> TyS<'tcx> {
...
-        Some(self.kind().cmp(other.kind()))
-    }
-}
+// To ensure correctness of incremental compilation,
+// `DefId` must not implement `Ord` or `PartialOrd`.
+// Since `TyKind` can contain a `DefId`, `TyS` cannot
+// implement `Ord` or `PartialOrd`.
+// See https://github.com/rust-lang/rust/issues/90317.
+impl !Ord for TyS<'_> {}
+impl !PartialOrd for TyS<'_> {}
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/mod.rs:426]: I don't think explicit negative impls are necessary here.

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 5]
```diff
@@ -1076,18 +1078,9 @@ pub trait PrettyPrinter<'tcx>:
         // Builtin bounds.
         // FIXME(eddyb) avoid printing twice (needed to ensure
         // that the auto traits are sorted *and* printed via cx).
-        let mut auto_traits: Vec<_> =
+        let mut auto_traits: FxIndexSet<_> =
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/print/pretty.rs:1081]: This should keep being a `Vec`. Below you can sort by just the def path str using `sort_by_key()`.

File: compiler/rustc_middle/src/query/mod.rs (hunk 3 above)
Comment (pierwill) [reply compiler/rustc_middle/src/query/mod.rs:1080]: Possibly... Here are the uses of `all_local_trait_impls` and `hir().trait_impls`:

`all_local_trait_impls`:

- [compiler/rustc_middle/src/hir/map/mod.rs#L552](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_middle/src/hir/map/mod.rs#L552)
- [compiler/rustc_middle/src/hir/mod.rs#L109](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_middle/src/hir/mod.rs#L109)
- [compiler/rustc_middle/src/query/mod.rs#L1080](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_middle/src/query/mod.rs#L1080)
- [compiler/rustc_typeck/src/coherence/mod.rs#L203](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_typeck/src/coherence/mod.rs#L203)
- [compiler/rustc_typeck/src/coherence/orphan.rs#L15](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_typeck/src/coherence/orphan.rs#L15)

`hir().trait_impls`:

- [compiler/rustc_infer/src/infer/error_reporting/nice_region_error/static_impl_trait.rs#L446](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_infer/src/infer/error_reporting/nice_region_error/static_impl_trait.rs#L446)
- [compiler/rustc_middle/src/ty/trait_def.rs#L222](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_middle/src/ty/trait_def.rs#L222)
- [compiler/rustc_typeck/src/coherence/builtin.rs#L41](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_typeck/src/coherence/builtin.rs#L41)
- [compiler/rustc_typeck/src/coherence/mod.rs#L189](https://github.com/pierwill/rust/blob/39fc8d73e8b7aa062d8da8f4fd0c7b70cae62d32/compiler/rustc_typeck/src/coherence/mod.rs#L189)

**Comment by pierwill** [comment]
@rustbot ready

I suspect that many of the errors are from removing the various calls to `sort` in https://github.com/rust-lang/rust/pull/90749/commits/b1c47c90e4ca9570d1b2e2e01cae2ea222f9c9a7. However, I'm not sure how best to go about diagnosing the problems.

In an earlier iteration, I tried just dropping in `IndexMap`s, but there were still tons of UI errors. It seemed like data was out of order and getting lost. A better approach might require looking at how these data types are constructed and where, in order to see if the order can be made deterministic, if not meaningful.

**Comment by pierwill** [comment]
@rustbot ready

I suspect that many of the errors are from removing the various calls to `sort` in https://github.com/rust-lang/rust/pull/90749/commits/b1c47c90e4ca9570d1b2e2e01cae2ea222f9c9a7. However, I'm not sure how best to go about diagnosing the problems.

In an earlier iteration, I tried just dropping in `IndexMap`s, but there were still tons of UI errors. It seemed like data was out of order and getting lost. A better approach might require looking at how these data types are constructed and where, in order to see if the order can be made deterministic, if not meaningful.

File: compiler/rustc_borrowck/src/region_infer/dump_mir.rs [hunk 6]
```diff
@@ -71,8 +71,8 @@ impl<'tcx> RegionInferenceContext<'tcx> {
             }
         }
 
-        let mut constraints: Vec<_> = self.constraints.outlives().iter().collect();
-        constraints.sort();
+        let constraints: Vec<_> = self.constraints.outlives().iter().collect();
+        // constraints.sort();
```
Comment (pierwill) [inline compiler/rustc_borrowck/src/region_infer/dump_mir.rs:75]: On a fresh branch off master, removing this sort causes UI errors.

File: compiler/rustc_borrowck/src/region_infer/mod.rs [hunk 7]
```diff
@@ -611,8 +611,8 @@ impl<'tcx> RegionInferenceContext<'tcx> {
     #[instrument(skip(self, _body), level = "debug")]
     fn propagate_constraints(&mut self, _body: &Body<'tcx>) {
         debug!("constraints={:#?}", {
-            let mut constraints: Vec<_> = self.constraints.outlives().iter().collect();
-            constraints.sort();
+            let constraints: Vec<_> = self.constraints.outlives().iter().collect();
```
Comment (pierwill) [inline compiler/rustc_borrowck/src/region_infer/mod.rs:614]: On a fresh branch off master, removing this sort causes UI errors.

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 8]
```diff
@@ -89,7 +89,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                  err: &mut DiagnosticBuilder<'_>,
                                  mut sources: Vec<CandidateSource>,
                                  sugg_span: Span| {
-            sources.sort();
+            // FIXME: cannot sort DefIds!
+            // sources.sort();
```
Comment (pierwill) [inline compiler/rustc_typeck/src/check/method/suggest.rs:93]: On a fresh branch off master, removing this sort causes UI errors.

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 9]
```diff
@@ -538,7 +539,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 })
                                 .collect::<Vec<_>>();
                             if !inherent_impls_candidate.is_empty() {
-                                inherent_impls_candidate.sort();
+                                // inherent_impls_candidate.sort();
```
Comment (pierwill) [inline compiler/rustc_typeck/src/check/method/suggest.rs:542]: On a fresh branch off master, removing this sort causes UI errors.

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 10]
```diff
@@ -1129,14 +1130,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-        let mut preds: Vec<_> = errors
+        let preds: Vec<_> = errors
             .iter()
             .filter_map(|e| match e.obligation.predicate.kind().skip_binder() {
                 ty::PredicateKind::Trait(pred) => Some(pred),
                 _ => None,
             })
             .collect();
-        preds.sort_by_key(|pred| (pred.def_id(), pred.self_ty()));
+        // preds.sort_by_key(|pred| (pred.def_id(), pred.self_ty()));
```
Comment (pierwill) [inline compiler/rustc_typeck/src/check/method/suggest.rs:1140]: On a fresh branch off master, removing this sort causes UI errors.

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 11]
```diff
@@ -1417,7 +1418,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     ) -> bool {
         if !valid_out_of_scope_traits.is_empty() {
             let mut candidates = valid_out_of_scope_traits;
-            candidates.sort();
+            // candidates.sort();
```
Comment (pierwill) [inline compiler/rustc_typeck/src/check/method/suggest.rs:1421]: On a fresh branch off master, removing this sort causes UI errors.

File: compiler/rustc_middle/src/ty/fold.rs [hunk 12]
```diff
@@ -744,12 +743,12 @@ impl<'tcx> TyCtxt<'tcx> {
         self,
         value: Binder<'tcx, T>,
         mut fld_r: F,
-    ) -> (T, BTreeMap<ty::BoundRegion, ty::Region<'tcx>>)
+    ) -> (T, FxHashMap<ty::BoundRegion, ty::Region<'tcx>>)
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/fold.rs:746]: Is this FxHashMap used anyhere?

File: compiler/rustc_middle/src/ty/fold.rs [hunk 13]
```diff
@@ -796,14 +795,14 @@ impl<'tcx> TyCtxt<'tcx> {
         mut fld_r: F,
         fld_t: G,
         fld_c: H,
-    ) -> (T, BTreeMap<ty::BoundRegion, ty::Region<'tcx>>)
+    ) -> (T, FxHashMap<ty::BoundRegion, ty::Region<'tcx>>)
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/fold.rs:798]: Does any use site rely on the iteration order?

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 14]
```diff
@@ -946,7 +945,10 @@ pub trait PrettyPrinter<'tcx>:
         &mut self,
         trait_ref: ty::PolyTraitRef<'tcx>,
         proj_ty: Option<(DefId, ty::Binder<'tcx, Ty<'tcx>>)>,
-        traits: &mut FxHashMap<ty::PolyTraitRef<'tcx>, BTreeMap<DefId, ty::Binder<'tcx, Ty<'tcx>>>>,
+        traits: &mut FxHashMap<
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/print/pretty.rs:948]: Where is this FxHashMap used? Does any code depend on the iteration order?

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 15]
```diff
@@ -946,7 +945,10 @@ pub trait PrettyPrinter<'tcx>:
...
+            ty::PolyTraitRef<'tcx>,
+            FxHashMap<DefId, ty::Binder<'tcx, Ty<'tcx>>>,
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/print/pretty.rs:950]: Same question?

File: compiler/rustc_middle/src/ty/subst.rs [hunk 16]
```diff
@@ -77,17 +76,17 @@ impl<'tcx> fmt::Debug for GenericArg<'tcx> {
...
+//     fn cmp(&self, other: &GenericArg<'_>) -> Ordering {
+//         self.unpack().cmp(&other.unpack())
+//     }
+// }
+
+// impl<'tcx> PartialOrd for GenericArg<'tcx> {
+//     fn partial_cmp(&self, other: &GenericArg<'_>) -> Option<Ordering> {
+//         Some(self.cmp(&other))
+//     }
+// }
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/subst.rs]: To be removed before approval.

File: compiler/rustc_middle/src/ty/adt.rs [hunk 17]
```diff
@@ -99,19 +98,19 @@ pub struct AdtDef {
...
+//     }
+// }
+
+// /// There should be only one AdtDef for each `did`, therefore
+// /// it is fine to implement `Ord` only based on `did`.
+// impl Ord for AdtDef {
+//     fn cmp(&self, other: &AdtDef) -> Ordering {
+//         self.did.cmp(&other.did)
+//     }
+// }
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/adt.rs]: To be removed before approval.

File: compiler/rustc_middle/src/query/mod.rs (hunk 3 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/query/mod.rs:1080]: If the iteration order matters, we may need to use an FxIndexMap and audit how these maps are built.

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs [hunk 18]
```diff
@@ -331,12 +331,8 @@ fn add_unused_functions<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>) {
     // Add a new `FunctionCoverage` to the `function_coverage_map`, with unreachable code regions
     // for each region in it's MIR.
 
-    // Convert the `HashSet` of `codegenned_def_ids` to a sortable vector, and sort them.
-    let mut sorted_codegenned_def_ids: Vec<DefId> = codegenned_def_ids.iter().copied().collect();
-    sorted_codegenned_def_ids.sort_unstable();
```
Comment (cjgillot) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:336]: Why was this sorted? Should the comment be corrected?

File: compiler/rustc_infer/src/infer/region_constraints/mod.rs [hunk 19]
```diff
@@ -90,7 +89,7 @@ pub type VarInfos = IndexVec<RegionVid, RegionVariableInfo>;
 pub struct RegionConstraintData<'tcx> {
     /// Constraints of the form `A <= B`, where either `A` or `B` can
     /// be a region variable (or neither, as it happens).
-    pub constraints: BTreeMap<Constraint<'tcx>, SubregionOrigin<'tcx>>,
+    pub constraints: FxHashMap<Constraint<'tcx>, SubregionOrigin<'tcx>>,
```
Comment (cjgillot) [inline compiler/rustc_infer/src/infer/region_constraints/mod.rs:92]: Does any code depend on the iteration order?

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 20]
```diff
@@ -513,9 +512,9 @@ pub struct BoundVarReplacer<'me, 'tcx> {
     // These three maps track the bound variable that were replaced by placeholders. It might be
     // nice to remove these since we already have the `kind` in the placeholder; we really just need
     // the `var` (but we *could* bring that into scope if we were to track them as we pass them).
-    mapped_regions: BTreeMap<ty::PlaceholderRegion, ty::BoundRegion>,
-    mapped_types: BTreeMap<ty::PlaceholderType, ty::BoundTy>,
-    mapped_consts: BTreeMap<ty::PlaceholderConst<'tcx>, ty::BoundVar>,
+    mapped_regions: FxHashMap<ty::PlaceholderRegion, ty::BoundRegion>,
+    mapped_types: FxHashMap<ty::PlaceholderType, ty::BoundTy>,
+    mapped_consts: FxHashMap<ty::PlaceholderConst<'tcx>, ty::BoundVar>,
```
Comment (cjgillot) [inline compiler/rustc_trait_selection/src/traits/project.rs:517]: Does any code depend on the iteration order?

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 21]
```diff
@@ -1916,8 +1916,8 @@ impl Ord for TraitInfo {
         // Local crates are more important than remote ones (local:
         // `cnum == 0`), and otherwise we throw in the defid for totality.
 
-        let lhs = (other.def_id.krate, other.def_id);
-        let rhs = (self.def_id.krate, self.def_id);
+        let (lhs, _) = (other.def_id.krate, other.def_id);
+        let (rhs, _) = (self.def_id.krate, self.def_id);
         lhs.cmp(&rhs)
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/check/method/suggest.rs:1920]: This impl is wrong and will probably produce surprising results.
We implement `Eq`, so we expect a real equality, not just `CrateNum`.
Is the `PartialOrd`/`Ord` bound used? Where? Can we live without?

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 22]
```diff
@@ -1338,7 +1337,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         }
 
         // Use a `BTreeSet` to keep output in a more consistent order.
-        let mut associated_types: FxHashMap<Span, BTreeSet<DefId>> = FxHashMap::default();
+        let mut associated_types: FxHashMap<Span, FxHashSet<DefId>> = FxHashMap::default();
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/astconv/mod.rs:1340]: Do we still rely on the iteration order?

File: compiler/rustc_typeck/src/outlives/utils.rs [hunk 23]
```diff
@@ -1,14 +1,14 @@
...
 use rustc_middle::ty::{self, Region, RegionKind, Ty, TyCtxt};
 use rustc_span::Span;
 use smallvec::smallvec;
-use std::collections::BTreeMap;
 
 /// Tracks the `T: 'a` or `'a: 'a` predicates that we have inferred
 /// must be added to the struct header.
 pub type RequiredPredicates<'tcx> =
-    BTreeMap<ty::OutlivesPredicate<GenericArg<'tcx>, ty::Region<'tcx>>, Span>;
+    FxHashMap<ty::OutlivesPredicate<GenericArg<'tcx>, ty::Region<'tcx>>, Span>;
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/outlives/utils.rs:11]: Do we rely somewhere on the iteration order?

**Review by cjgillot — CHANGES_REQUESTED** [review]
Thanks for the PR @pierwill, I expected much less use of these ordering impls.

Could you please reorder/remix your commits into commits that pass `x.py check`? This makes it much easier to review and to better see the consequences of each modification.
The main concern is the iteration order in each place where we replace a `BTree` by a hash-map. 
~You can use the internal lint `rustc::potential_query_instability` to help finding such issues.~ (lint has been removed)

File: compiler/rustc_middle/src/ty/fold.rs (hunk 12 above)
Comment (pierwill) [reply compiler/rustc_middle/src/ty/fold.rs:746]: Actually, it might not be! 👀

Comment (pierwill) [reply compiler/rustc_middle/src/ty/fold.rs:746]: Maybe not? 👀 🤔 I'm not entirely sure how to check. The `lsp-find-references` command in Emacs shows several places where the returned map is ignored:

```
compiler/rustc_borrowck/src/universal_regions.rs
736:         let (value, _map) = self.tcx.replace_late_bound_regions(value, |br| {
compiler/rustc_middle/src/ty/fold.rs
821:         self.replace_late_bound_regions(value, |br| {
911:         self.replace_late_bound_regions(value, |_| self.lifetimes.re_erased).0
928:             .replace_late_bound_regions(sig, |_| {
compiler/rustc_middle/src/ty/print/pretty.rs
2160:             self.tcx.replace_late_bound_regions(value.clone(), |br| {
compiler/rustc_typeck/src/collect.rs
457:                                             .replace_late_bound_regions(poly_trait_ref, |_| {
```

File: compiler/rustc_middle/src/ty/fold.rs (hunk 13 above)
Comment (pierwill) [reply compiler/rustc_middle/src/ty/fold.rs:798]: Maybe here?  https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_borrowck/src/type_check/mod.rs#L1589

File: compiler/rustc_typeck/src/astconv/mod.rs (hunk 22 above)
Comment (pierwill) [reply compiler/rustc_typeck/src/astconv/mod.rs:1340]: Yes: https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_typeck/src/astconv/mod.rs#L1403

File: compiler/rustc_typeck/src/outlives/utils.rs (hunk 23 above)
Comment (pierwill) [reply compiler/rustc_typeck/src/outlives/utils.rs:11]: Yes: https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_typeck/src/outlives/explicit.rs#L31

File: compiler/rustc_infer/src/infer/region_constraints/mod.rs (hunk 19 above)
Comment (pierwill) [reply compiler/rustc_infer/src/infer/region_constraints/mod.rs:92]: Yes:

- https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs#L189
- https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs#L229
- https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs#L449
- https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs#L618
- https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_trait_selection/src/traits/auto_trait.rs#L515

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs (hunk 18 above)
Comment (pierwill) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:336]: I'm not sure why it was sorted before. We'll eventually need to remove this, right?

File: compiler/rustc_trait_selection/src/traits/project.rs (hunk 20 above)
Comment (pierwill) [reply compiler/rustc_trait_selection/src/traits/project.rs:517]: Possibly here?

https://github.com/rust-lang/rust/blob/f1ce0e6a00593493a12e0e3662119786c761f375/compiler/rustc_trait_selection/src/traits/project.rs#L464-L484

File: compiler/rustc_middle/src/ty/print/pretty.rs (hunk 14 above)
Comment (pierwill) [reply compiler/rustc_middle/src/ty/print/pretty.rs:948]: This map is used in two places:

- https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_middle/src/ty/print/pretty.rs#L797
- https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_middle/src/ty/print/pretty.rs#L806-L807

File: compiler/rustc_middle/src/ty/print/pretty.rs (hunk 15 above)
Comment (pierwill) [reply compiler/rustc_middle/src/ty/print/pretty.rs:950]: Used in the same calls as above https://github.com/rust-lang/rust/pull/90749#discussion_r779062831.

File: compiler/rustc_typeck/src/check/method/suggest.rs (hunk 21 above)
Comment (pierwill) [reply compiler/rustc_typeck/src/check/method/suggest.rs]: The ordering is used here https://github.com/rust-lang/rust/blob/088dd7cb54f57bd1c870a343cb5dc3d3d2703761/compiler/rustc_typeck/src/check/method/suggest.rs#L1651 and in the `PartialEq` implementation: https://github.com/rust-lang/rust/blob/088dd7cb54f57bd1c870a343cb5dc3d3d2703761/compiler/rustc_typeck/src/check/method/suggest.rs#L1906

Comment (pierwill) [reply compiler/rustc_typeck/src/check/method/suggest.rs]: And the `Eq`,`PartialEq` are currently used for deduplication: https://github.com/rust-lang/rust/blob/088dd7cb54f57bd1c870a343cb5dc3d3d2703761/compiler/rustc_typeck/src/check/method/suggest.rs#L1652

Comment (pierwill) [reply compiler/rustc_typeck/src/check/method/suggest.rs]: Would it work or defeat our purpose to sort the TraitInfos by DefIndex?

**Comment by pierwill** [comment]
Closing this (for now), as I'm looking at making this change in a series of smaller PRs.

**Pull request closed without merging**